### ISSUE TITLE
Action exceptions should be recorded as a part of blockchain

### ISSUE THREAD

**Issue by dahlia** [planetarium/libplanet#860]
Due to the limitation of the current `IAction` API, validation logic tends to be duplicated (one in the `Execute()` method and other one in the game UI code).  If exceptions raised by actions are recorded as a part of blockchain, such duplicated validation code in the game UI side can be replaced by reusing the result/error code of an action.

**Comment by dahlia** [comment]
@ipdae Any ideas?

**Comment by ipdae** [comment]
In my opinion, it would be nice to be able to retrieve the exception that occurs when executing the action in the game through an interface such as `EvaluationOuputStates.Exception`.

**Comment by longfin** [comment]
> In my opinion, it would be nice to be able to retrieve the exception that occurs when executing the action in the game through an interface such as EvaluationOuputStates.Exception.

`ActionEvaluation` is public class but the game using Libplanet doesn't access it directly. Instead, the game accesses action's results by IAction.Render().

so there are several options if we want to allow to access generated exception to users.

- give whole `ActionEvaluation` as `.Render()'s argument instead of `IActionContext` and `IAccountStateDelta`.
- add `exception` to `.Render()`'s arguments.
- attach it to `IAccountStateDelta` or `IActionContext`.

The former is more natural to me. Is there any idea? @planetarium/libplanet @ipdae

**Comment by dahlia** [comment]
How about add a new distinct method to `IAction` besides `Render()`/`UnrenderError()`, e.g., `RenderError()`/`UnrenderError()`?  `IAction` would look like:

```cs
public interface IAction
{
    IAccountStateDelta Execute(IActionContext context);
    void Render(IActionContext context, IAccountStateDelta nextStates);
    void Unrender(IActionContext context, IAccountStateDelta nextStates);
    void RenderError(IActionContext context, Exception exception);    // 🆕
    void UnrenderError(IActionContext context, Exception exception);  // 🆕
    IValue PlainValue { get; }
    void LoadPlainValue(IValue plainValue);
}
```

**Comment by longfin** [comment]
> How about adding a new distinct method to `IAction` besides `Render()`/`Unrender()`, e.g., `RenderError()`/`UnrenderError()`? `IAction` would look like:
> 
> ```cs
> public interface IAction
> {
>     IAccountStateDelta Execute(IActionContext context);
>     void Render(IActionContext context, IAccountStateDelta nextStates);
>     void Unrender(IActionContext context, IAccountStateDelta nextStates);
>     void RenderError(IActionContext context, Exception exception);    // 🆕
>     void UnrenderError(IActionContext context, Exception exception);  // 🆕
>     IValue PlainValue { get; }
>     void LoadPlainValue(IValue plainValue);
> }
> ```

`RenderError()` / `UnrenderError()` seem good. I'll take them.

**Comment by longfin** [comment]
It seems resolved by #875
